test-data/unit/check-narrowing.test

```diff
@@ -1,3 +1,86 @@
+[case testNarrowingParentWithStrsBasic]
+from dataclasses import dataclass
+from typing import NamedTuple, Tuple, Union
+from typing_extensions import Literal, TypedDict
+
+class Object1:
+    key: Literal["A"]
+    foo: int
+class Object2:
+    key: Literal["B"]
+    bar: str
+
+@dataclass
+class Dataclass1:
+    key: Literal["A"]
+    foo: int
+@dataclass
+class Dataclass2:
+    key: Literal["B"]
+    foo: str
+
+class NamedTuple1(NamedTuple):
+    key: Literal["A"]
+    foo: int
+class NamedTuple2(NamedTuple):
+    key: Literal["B"]
+    foo: str
+
+Tuple1 = Tuple[Literal["A"], int]
+Tuple2 = Tuple[Literal["B"], str]
+
+class TypedDict1(TypedDict):
+    key: Literal["A"]
+    foo: int
+class TypedDict2(TypedDict):
+    key: Literal["B"]
+    foo: str
+
+x1: Union[Object1, Object2]
+if x1.key == "A":
+    reveal_type(x1)         # N: Revealed type is '__main__.Object1'
+    reveal_type(x1.key)     # N: Revealed type is 'Literal['A']'
+else:
+    reveal_type(x1)         # N: Revealed type is '__main__.Object2'
+    reveal_type(x1.key)     # N: Revealed type is 'Literal['B']'
+
+x2: Union[Dataclass1, Dataclass2]
+if x2.key == "A":
+    reveal_type(x2)         # N: Revealed type is '__main__.Dataclass1'
+    reveal_type(x2.key)     # N: Revealed type is 'Literal['A']'
+else:
+    reveal_type(x2)         # N: Revealed type is '__main__.Dataclass2'
+    reveal_type(x2.key)     # N: Revealed type is 'Literal['B']'
+
+x3: Union[NamedTuple1, NamedTuple2]
+if x3.key == "A":
+    reveal_type(x3)         # N: Revealed type is 'Tuple[Literal['A'], builtins.int, fallback=__main__.NamedTuple1]'
+    reveal_type(x3.key)     # N: Revealed type is 'Literal['A']'
+else:
+    reveal_type(x3)         # N: Revealed type is 'Tuple[Literal['B'], builtins.str, fallback=__main__.NamedTuple2]'
+    reveal_type(x3.key)     # N: Revealed type is 'Literal['B']'
+if x3[0] == "A":
+    reveal_type(x3)         # N: Revealed type is 'Tuple[Literal['A'], builtins.int, fallback=__main__.NamedTuple1]'
+    reveal_type(x3[0])      # N: Revealed type is 'Literal['A']'
+else:
+    reveal_type(x3)         # N: Revealed type is 'Tuple[Literal['B'], builtins.str, fallback=__main__.NamedTuple2]'
+    reveal_type(x3[0])      # N: Revealed type is 'Literal['B']'
+
+x4: Union[Tuple1, Tuple2]
+if x4[0] == "A":
+    reveal_type(x4)         # N: Revealed type is 'Tuple[Literal['A'], builtins.int]'
+    reveal_type(x4[0])      # N: Revealed type is 'Literal['A']'
+else:
+    reveal_type(x4)         # N: Revealed type is 'Tuple[Literal['B'], builtins.str]'
+    reveal_type(x4[0])      # N: Revealed type is 'Literal['B']'
+
+x5: Union[TypedDict1, TypedDict2]
+if x5["key"] == "A":
+    reveal_type(x5)         # N: Revealed type is 'TypedDict('__main__.TypedDict1', {'key': Literal['A'], 'foo': builtins.int})'
+else:
+    reveal_type(x5)         # N: Revealed type is 'TypedDict('__main__.TypedDict2', {'key': Literal['B'], 'foo': builtins.str})'
+[builtins fixtures/primitives.pyi]
```

Comment:
Maybe add couple more tests with typed dicts (for example nested typed dicts), since this is pretty important use case.

Reply:
Forgot to mention this earlier, but I added a few more tests, mostly copying some of the existing ones I had for enums and objects. LMK if there are more test cases you'd like to see.